### PR TITLE
fix: reject calls to controllers when alarms are raised

### PR DESCRIPTION
**Pull request by InoMurko**
/

## Overview

If ethereum client connectivity is down, prevent calls going further into the stack.
If monitor looses all it's children, prevent calls going further into the stack.

## Changes

Describe your changes and implementation choices. More details make PRs easier to review.

- further simplified Monitors
- unified setting alarms
- if a monitor looses all it's children it'll raise an alarm _chain crash_

## Testing

Unit tests

Child Chain state is indeed not coupled with Eth. So I've removed the plug check for now and see how it behaves. 

Screen Shot 2019-10-02 at 22 15 41

Screen Shot 2019-10-02 at 22 15 41

### CODE REVIEW

**Comment by InoMurko** [comment]
I think what confused me is that Watcher's OMG.State is coupled with Ethereum connectivity.
Screen Shot 2019-10-02 at 22 34 37

File: apps/omg_status/lib/omg_status/metric/statsd_monitor.ex [hunk 1]
```diff
@@ -120,10 +120,10 @@ defmodule OMG.Status.Metric.StatsdMonitor do
   defp raise_clear(_alarm_module, true, false), do: :ok
 
   defp raise_clear(alarm_module, false, false),
-    do: alarm_module.set({:statsd_client_connection, Node.self(), __MODULE__})
+    do: alarm_module.set(alarm_module.statsd_client_connection(__MODULE__))
```
Comment (pnowosie) [inline apps/omg_status/lib/omg_status/metric/statsd_monitor.ex]: `alarm_module` could be omitted at `init()`. Let's replace:
`alarm_module = Keyword.get(opts, :alarm_module)` with: 
`alarm_module = Keyword.fetch!(opts, :alarm_module)`

Comment (pnowosie) [reply apps/omg_status/lib/omg_status/metric/statsd_monitor.ex]: I guess `child_module` could be made required the same way

File: apps/omg_child_chain_rpc/lib/omg_child_chain_rpc/web/controllers/block.ex [hunk 2]
```diff
@@ -18,7 +18,8 @@ defmodule OMG.ChildChainRPC.Web.Controller.Block do
   """
 
   use OMG.ChildChainRPC.Web, :controller
-  plug(OMG.ChildChainRPC.Plugs.Health)
+  # check for health before calling action
+  # plug(OMG.ChildChainRPC.Plugs.Health)
```
Comment (pnowosie) [inline apps/omg_child_chain_rpc/lib/omg_child_chain_rpc/web/controllers/block.ex]: Was you intention to leave it commented? Consider delete lines or uncomment.

File: apps/omg_child_chain_rpc/lib/omg_child_chain_rpc/web/controllers/transaction.ex [hunk 3]
```diff
@@ -18,7 +18,8 @@ defmodule OMG.ChildChainRPC.Web.Controller.Transaction do
   """
 
   use OMG.ChildChainRPC.Web, :controller
-  plug(OMG.ChildChainRPC.Plugs.Health)
+  # check for health before calling action
+  # plug(OMG.ChildChainRPC.Plugs.Health)
```
Comment (pnowosie) [inline apps/omg_child_chain_rpc/lib/omg_child_chain_rpc/web/controllers/transaction.ex]: Was you intention to leave it commented?

File: apps/omg_child_chain_rpc/lib/omg_child_chain_rpc/web/controllers/block.ex (hunk 2 above)
Comment (InoMurko) [reply apps/omg_child_chain_rpc/lib/omg_child_chain_rpc/web/controllers/block.ex]: it used to be enabled, so I commented it out so that we could have a discussion on what to do, cc @pdobacz

File: apps/omg_child_chain_rpc/lib/omg_child_chain_rpc/web/controllers/transaction.ex (hunk 3 above)
Comment (InoMurko) [reply apps/omg_child_chain_rpc/lib/omg_child_chain_rpc/web/controllers/transaction.ex]: it used to be enabled, so I commented it out so that we could have a discussion on what to do, cc @pdobacz

File: apps/omg_status/lib/omg_status/metric/statsd_monitor.ex (hunk 1 above)
Comment (InoMurko) [reply apps/omg_status/lib/omg_status/metric/statsd_monitor.ex]: I'm not exactly sure I understand what you mean by this comment... could you elaborate pls?

File: apps/omg_child_chain_rpc/lib/omg_child_chain_rpc/web/controllers/block.ex (hunk 2 above)
Comment (pdobacz) [reply apps/omg_child_chain_rpc/lib/omg_child_chain_rpc/web/controllers/block.ex]: AFAIK, eth conn isn't required for the chch's `block.get` and `transaction.submit` (at least not to successfully process the request per se, as we should always know and react if eth conn is lost!).

Judging by your comment, you're intending to go with these checks removed and I think let's do it (and I guess let's remove the commented code x2)

File: apps/omg_status/lib/omg_status/metric/statsd_monitor.ex (hunk 1 above)
Comment (pdobacz) [reply apps/omg_status/lib/omg_status/metric/statsd_monitor.ex]: Jumping in (Papa is out). I think this is meant: https://github.com/omisego/elixir-omg/blob/8795fb527d3eb4c347142fb4e1b26faa7f380ff3/apps/omg_status/lib/omg_status/metric/statsd_monitor.ex#L48

`get` is used there, so this can be `nil` and the suggestion is to `fetch!` there to avoid a `nil.statsd_client_connection` and fail fast instead.

File: apps/omg_status/lib/status.ex [hunk 4]
```diff
@@ -23,7 +23,7 @@ defmodule OMG.Status do
   # if ETS table has a tuple entry in form of {:boot_in_progress, 0}, return false
   # if ETS table has a tuple entry in form of {:boot_in_progress, 1}, return true
   @health_match List.flatten(
-                  for n <- [:boot_in_progress, :ethereum_client_connection],
+                  for n <- [:boot_in_progress, :ethereum_client_connection, :chain_crash],
```
Comment (pdobacz) [inline apps/omg_status/lib/status.ex]: `:chain_crash` isn't too clear. I'd also associate a total `crash` with sth leading BEAM to exiting, which is not the case here.

Also (and more importantly) this alarm pertains to both watcher and child chain, whereas `chain` is quite generic and abstract. Even if the watcher is in `chain_crash`, one shouldn't think that the chain crashed or stopped since it is might be just fine (both the `child chain` and the `root chain`.

Alternatives: `main_supervisor_exited`, `ethereum_supervisor_exited/halted`, `monitor_exited`. I think we can be more specific here.

File: apps/omg_child_chain_rpc/lib/omg_child_chain_rpc/web/controllers/block.ex (hunk 2 above)
Comment (InoMurko) [reply apps/omg_child_chain_rpc/lib/omg_child_chain_rpc/web/controllers/block.ex]: I would disagree. The service is not functioning properly at the fundamental level - we don't have ethereum connectivity - and we have no idea for how long. Do we really want to introduce state changes when we don't know how long the outage is?

Comment (pdobacz) [reply apps/omg_child_chain_rpc/lib/omg_child_chain_rpc/web/controllers/block.ex]: Not sure what you're disagreeing with. I didn't really write anything decisive, if you want to block the requests then uncomment, if you want to let them through - remove.

If we prefer to err on the safe side and block even for transient losses of eth-conn, let's uncomment.

If we're confident that via alerts we'll learn that eth-conn's lost and react before it becomes an issue (or stop the services if need be), let's remove.

Probably taking either path would warrant opening an issue to follow-up.

**Comment by InoMurko** [comment]
hey @pdobacz I've remembered I have this PR open which kinda addresses the issue https://github.com/omisego/elixir-omg/issues/1141 as well. can you take a look and let me know what you think. AlarmPrinter doesn't have tests yet. everything else should be solid. let me know your thoughts and if this is enough to close #1141

**Comment by InoMurko** [comment]
Let me add that the next step is to duplicate the loudness by pushing it to Datadog as well.

**Comment by pdobacz** [comment]
I think as long as suitable loudness of failures/alarms is guaranteed (ideally a crash in unrecoverable, critical errors that won't resolve without developer's intervention, as discussed #1141) this approach LGTM.

I think we did slightly disagree on the "suitable loudness" part, but to not block, let's do it as you propose and promise ourselves to fix forward, if we run into a situation where loudness was to little (and consider crashing then).

Last thing: I can't read out of the code whether this addresses #1141 in fact. Can you please run the `:observer.start` test as #1141 mentions to double check (possibly killing various other processes in the tree as well)?

(oh and let's rethink the `chain_crash` name as my other comment)

**Comment by InoMurko** [comment]
I've mentioned I've lost a few commits that add a few important bits... will ping you when it's ready!

**Comment by InoMurko** [comment]
A bit of background on this PR and how it was designed:

Screen Shot 2019-12-24 at 11 45 27

This is how the supervision tree of the most important bit  in Child Chain (and similarly Watcher) looks like.
OMG.ChildChain.Monitor process is hooked on alarms related to ethereum connection (clients - infura), in case of hickups alarms will be raised and the Monitor is going to be able to react (restart it's underlying SyncSupervisor). All children of Sync Supervisor need a stable connection to ethereum clients and they will die if they can't RPC the ethereum nodes.

But, that's not the only reason they could die. They could die because of bugs or other implementation mistakes we're not able to digest. That WILL cause a complete outage and restarts will probably not help with that. For restarts, we still have the supervisor mechanism of  the SyncSupervisor, but once this supervisor does it's thing and can't help it's children anymore, it'll shutdown. At this point, the `OMG.ChildChain.Monitor` is going to receive a message and raise an alarm `main_supervisor_halted`. This will prevent any further requests to reach the node internals.
```
2019-12-24 11:41:50.168 [error] module=OMG.ChildChain.Monitor function=handle_info/2 ⋅Child Chain supervisor crashed. Raising alarm. Reason :killed⋅
2019-12-24 11:41:50.169 [info] module=OMG.ChildChain.Monitor function=handle_event/2 ⋅Monitor got event: {:set_alarm, {:main_supervisor_halted, %{node: :"child_chain@127.0.0.1", reporter: OMG.ChildChain.Monitor}}}. Ignoring.⋅
2019-12-24 11:41:50.169 [info] module=OMG.Eth.EthereumClientMonitor function=handle_event/2 ⋅Eth client monitor got event: {:set_alarm, {:main_supervisor_halted, %{node: :"child_chain@127.0.0.1", reporter: OMG.ChildChain.Monitor}}}. Ignoring.⋅
```
SyncSupervisor children got a new child called OMG.ChildChain.ChildManager (or `OMG.Watcher.ChildManager`). It's purpose is to be the last child in the list of children. Last one because it's purpose is to report a successful boot of all Ethereum Event listeners and other important children of the SyncSupervisor. It will send a message it the Monitor and shut down itself (`use GenServer, restart: :transient`).  

`OMG.Status.AlarmPrinter` has been added to omg status application. The main purpose of the process is to print all the raised alarms every once in a while (implemented backoff). That will give us a certain level of loudness. These alarms will later get hooked to Datadog + Slack.
```
2019-12-24 11:43:22.040 [warn] module=OMG.Status.AlarmPrinter function=handle_info/2 ⋅An alarm was raised {:main_supervisor_halted, %{node: :"child_chain@127.0.0.1", reporter: OMG.ChildChain.Monitor}}⋅
2019-12-24 11:43:22.040 [warn] module=OMG.Status.AlarmPrinter function=handle_info/2 ⋅An alarm was raised {:system_memory_high_watermark, []}⋅
2019-12-24 11:43:22.040 [warn] module=OMG.Status.AlarmPrinter function=handle_info/2 ⋅An alarm was raised {{:disk_almost_full, '/System/Volumes/Data'}, []}⋅
```
------------------------------------------------
What happens if Ethereum client connection drops:
```
make init_test
make start-services
```
```
make start-child_chain
```
```
docker ps
docker stop <geth container id>
```
```
2019-12-24 12:00:45.858 [info] module=OMG.ChildChain.Monitor function=handle_event/2 ⋅Monitor got event: {:set_alarm, {:ethereum_client_connection, %{node: :"child_chain@127.0.0.1", reporter: OMG.Eth.EthereumClientMonitor}}}. Ignoring.⋅
2019-12-24 12:00:45.858 [warn] module=OMG.Eth.EthereumClientMonitor function=handle_event/2 ⋅Health check raised :ethereum_client_connection alarm.⋅
```

All Sync Supervisor children begin to crash and the Sync Supervisor is trying to restart them `:one_for_one` + :max_restarts 3, :max_seconds 5. 
```
2019-12-24 12:00:46.321 [error] module=gen_server function=error_info/7 ⋅GenServer OMG.ChildChain.BlockQueue.Server terminating
** (MatchError) no match of right hand side value: {:error, :econnrefused}
    (omg_child_chain) lib/omg_child_chain/block_queue.ex:138: OMG.ChildChain.BlockQueue.Server.handle_info/2
    (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:711: :gen_server.handle_msg/6
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message: :check_ethereum_status⋅
2019-12-24 12:00:46.321 [info] module=OMG.ChildChain.BlockQueue.Server function=handle_continue/2 ⋅Starting Elixir.OMG.ChildChain.BlockQueue.Server service.⋅
2019-12-24 12:00:46.323 [error] module=gen_server function=error_info/7 ⋅GenServer OMG.ChildChain.BlockQueue.Server terminating
** (MatchError) no match of right hand side value: {:error, :geth_not_listening}
    (omg_child_chain) lib/omg_child_chain/block_queue.ex:57: OMG.ChildChain.BlockQueue.Server.handle_continue/2
    (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:388: :gen_server.loop/7
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message: {:continue, :setup}⋅
2019-12-24 12:00:46.323 [info] module=OMG.ChildChain.BlockQueue.Server function=handle_continue/2 ⋅Starting Elixir.OMG.ChildChain.BlockQueue.Server service.⋅
2019-12-24 12:00:46.325 [error] module=gen_server function=error_info/7 ⋅GenServer OMG.ChildChain.BlockQueue.Server terminating
** (MatchError) no match of right hand side value: {:error, :geth_not_listening}
    (omg_child_chain) lib/omg_child_chain/block_queue.ex:57: OMG.ChildChain.BlockQueue.Server.handle_continue/2
    (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:388: :gen_server.loop/7
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message: {:continue, :setup}⋅
2019-12-24 12:00:46.325 [info] module=OMG.ChildChain.BlockQueue.Server function=handle_continue/2 ⋅Starting Elixir.OMG.ChildChain.BlockQueue.Server service.⋅
2019-12-24 12:00:46.326 [error] module=gen_server function=error_info/7 ⋅GenServer OMG.ChildChain.BlockQueue.Server terminating
** (MatchError) no match of right hand side value: {:error, :geth_not_listening}
    (omg_child_chain) lib/omg_child_chain/block_queue.ex:57: OMG.ChildChain.BlockQueue.Server.handle_continue/2
    (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:388: :gen_server.loop/7
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message: {:continue, :setup}⋅
```
After it reaches it's threshold...
```
2019-12-24 12:00:46.326 [error] module=OMG.ChildChain.Monitor function=handle_info/2 ⋅Child Chain supervisor crashed. Raising alarm. Reason :shutdown⋅
```
Screen Shot 2019-12-24 at 12 30 31

Alarm Printer is working:
```
2019-12-24 12:00:59.502 [warn] module=OMG.Status.AlarmPrinter function=handle_info/2 ⋅An alarm was raised {:main_supervisor_halted, %{node: :"child_chain@127.0.0.1", reporter: OMG.ChildChain.Monitor}}⋅
2019-12-24 12:00:59.502 [warn] module=OMG.Status.AlarmPrinter function=handle_info/2 ⋅An alarm was raised {:ethereum_client_connection, %{node: :"child_chain@127.0.0.1", reporter: OMG.Eth.EthereumClientMonitor}}⋅
2019-12-24 12:00:59.502 [warn] module=OMG.Status.AlarmPrinter function=handle_info/2 ⋅An alarm was raised {:system_memory_high_watermark, []}⋅
2019-12-24 12:00:59.502 [warn] module=OMG.Status.AlarmPrinter function=handle_info/2 ⋅An alarm was raised {{:disk_almost_full, '/System/Volumes/Data'}, []}⋅
```

If we bring geth back (read Infura):
```
2019-12-24 12:31:56.403 [warn] module=OMG.Eth.EthereumClientMonitor function=handle_event/2 ⋅Health check established connection to the client. :ethereum_client_connection alarm clearead.⋅
2019-12-24 12:31:56.404 [info] module=OMG.ChildChain.Monitor function=handle_cast/2 ⋅Monitor is restarting child %OMG.ChildChain.Monitor.Child{pid: #PID<0.1873.0>, spec: %{id: OMG.ChildChain.SyncSupervisor, restart: :permanent, start: {OMG.ChildChain.SyncSupervisor, :start_link, []}, type: :supervisor}}.⋅
2019-12-24 12:31:56.404 [info] module=OMG.ChildChain.SyncSupervisor function=init/1 ⋅Starting OMG.ChildChain.SyncSupervisor⋅
```
which reads like: the ethereum connection was established bringing the ⋅:ethereum_client_connection alarm down. Monitor got that event and started restating `SyncSupervisor`. Which in turn started restarting all it's children again:
```
2019-12-24 12:31:56.404 [info] module=OMG.ChildChain.BlockQueue.Server function=handle_continue/2 ⋅Starting Elixir.OMG.ChildChain.BlockQueue.Server service.⋅
2019-12-24 12:31:56.404 [info] module=OMG.RootChainCoordinator function=handle_continue/2 ⋅Starting Elixir.OMG.RootChainCoordinator service.⋅
2019-12-24 12:31:56.405 [info] module=OMG.EthereumEventListener function=handle_continue/2 ⋅Starting OMG.EthereumEventListener for depositor.⋅
2019-12-24 12:31:56.405 [info] module=OMG.EthereumEventListener function=handle_continue/2 ⋅Starting OMG.EthereumEventListener for in_flight_exit.⋅
2019-12-24 12:31:56.405 [info] module=OMG.EthereumEventListener function=handle_continue/2 ⋅Starting OMG.EthereumEventListener for piggyback.⋅
2019-12-24 12:31:56.405 [info] module=OMG.EthereumEventListener function=handle_continue/2 ⋅Starting OMG.EthereumEventListener for exiter.⋅
2019-12-24 12:31:56.406 [info] module=OMG.RootChainCoordinator function=handle_continue/2 ⋅Started OMG.RootChainCoordinator⋅
2019-12-24 12:31:56.420 [info] module=OMG.EthereumEventListener function=handle_continue/2 ⋅Started OMG.EthereumEventListener for piggyback, synced_height: 7052⋅
2019-12-24 12:31:56.424 [info] module=OMG.EthereumEventListener function=handle_continue/2 ⋅Started OMG.EthereumEventListener for depositor, synced_height: 7053⋅
2019-12-24 12:31:56.424 [info] module=OMG.EthereumEventListener function=handle_continue/2 ⋅Started OMG.EthereumEventListener for in_flight_exit, synced_height: 7053⋅
2019-12-24 12:31:56.424 [info] module=OMG.EthereumEventListener function=handle_continue/2 ⋅Started OMG.EthereumEventListener for exiter, synced_height: 7053⋅
2019-12-24 12:31:56.425 [info] module=OMG.ChildChain.BlockQueue.Server function=handle_continue/2 ⋅Starting BlockQueue at parent_height: 1792, parent_start: 67, mined_child_block: 0, stored_child_top_block: 1000⋅
2019-12-24 12:31:56.430 [info] module=OMG.ChildChain.BlockQueue.Server function=handle_continue/2 ⋅Starting BlockQueue, top_mined_hash: "0x0000000000000000000000000000000000000000000000000000000000000000"⋅
2019-12-24 12:31:56.430 [info] module=OMG.ChildChain.BlockQueue.Core function=enqueue_existing_blocks/3 ⋅Loaded with [] mined and [<<127, 85, 89, 87, 208, 9, 204, 165, 106, 201, 255, 118, 189, 138, 229, 101, 131, 92, 35, 108, 224, 252, 251, 216, 70, 118, 80, 89, 59, 198, 221, 72>>] enqueued⋅
2019-12-24 12:31:56.430 [info] module=OMG.ChildChain.BlockQueue.Server function=handle_continue/2 ⋅Started OMG.ChildChain.BlockQueue.Server⋅
2019-12-24 12:32:01.408 [info] module=OMG.ChildChain.Monitor function=handle_cast/2 ⋅Got a health checkin... clearing alarm main_supervisor_halted.⋅
``` 
Manager reported it's health which caused the alarms to clear:
```
2019-12-24 12:32:01.409 [info] module=OMG.ChildChain.Monitor function=handle_event/2 ⋅Monitor got event: {:clear_alarm, {:main_supervisor_halted, %{node: :"child_chain@127.0.0.1", reporter: OMG.ChildChain.Monitor}}}. Ignoring.⋅
2019-12-24 12:32:01.409 [info] module=OMG.Eth.EthereumClientMonitor function=handle_event/2 ⋅Eth client monitor got event: {:clear_alarm, {:main_supervisor_halted, %{node: :"child_chain@127.0.0.1", reporter: OMG.ChildChain.Monitor}}}. Ignoring.⋅
```

Cheers for reading! :)

File: apps/omg_child_chain/lib/omg_child_chain/fee_server.ex [hunk 5]
```diff
@@ -75,14 +75,12 @@ defmodule OMG.ChildChain.FeeServer do
           _ = Logger.warn("Fees are ignored. Updates have no effect.")
 
         _ ->
-          alarm = {:invalid_fee_file, Node.self(), __MODULE__}
-
           case update_fee_specs(state[:fee_adapter]) do
             :ok ->
-              Alarm.clear(alarm)
+              Alarm.clear(Alarm.invalid_fee_file(__MODULE__))
```
Comment (unnawut) [inline apps/omg_child_chain/lib/omg_child_chain/fee_server.ex]: Interesting pattern. I'm taking note of.

File: apps/omg_child_chain/lib/omg_child_chain/child_manager.ex [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+  require Logger
+  @timer 5_000
+  def start_link(args) do
+    GenServer.start_link(__MODULE__, args, name: __MODULE__)
+  end
+
+  def init(args) do
+    monitor = Keyword.fetch!(args, :monitor)
+    {:ok, _tref} = :timer.send_after(@timer, :health_checkin)
```
Comment (unnawut) [inline apps/omg_child_chain/lib/omg_child_chain/child_manager.ex]: Curious why we need to delay the checkin, and delay by 5s?

File: apps/omg_child_chain/test/omg_child_chain/integration/fixtures.exs [hunk 7]
```diff
@@ -96,4 +98,15 @@ defmodule OMG.ChildChain.Integration.Fixtures do
 
     {deposit_blknum, token_deposit_blknum}
   end
+
+  defp wait_for_web() do
```
Comment (unnawut) [inline apps/omg_child_chain/test/omg_child_chain/integration/fixtures.exs]: Is there an overall timeout? From the look it could leave the test stalling indefinitely?

Same in other modules.

File: apps/omg_db/test/omg_db/release_tasks/set_key_value_db_test.exs [hunk 8]
```diff
@@ -40,11 +40,11 @@ defmodule OMG.DB.ReleaseTasks.SetKeyValueDBTest do
     :ok = System.put_env("DB_PATH", "/tmp/YOLO/")
     :ok = SetKeyValueDB.init([])
     configuration = Enum.sort(Application.get_all_env(@app))
-    "/tmp/YOLO/watcher" = configuration[:path]
+    "/tmp/YOLO/watcher_info" = configuration[:path]
```
Comment (unnawut) [inline apps/omg_db/test/omg_db/release_tasks/set_key_value_db_test.exs]: Why the change? Just in case if there was an issue I should beware of.

File: apps/omg_status/test/omg_status/alert/alarm_printer_test.exs [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      %{previous_backoff: backoff} = :sys.get_state(alarm_printer)
+      :print_alarms = send(alarm_printer, :print_alarms)
+      min = backoff
+      max = backoff * 3
+      %{previous_backoff: new_backoff} = :sys.get_state(alarm_printer)
+      assert min < new_backoff and new_backoff < max
+    end)
+  end
+
+  defp find_warn_print(alarm_printer, alarms) do
```
Comment (unnawut) [inline apps/omg_status/test/omg_status/alert/alarm_printer_test.exs]: `find_warn_print` would make a great techno-thriller movie name! 🎥 🚀

File: apps/omg_watcher_info/mix.exs [hunk 10]
```diff
@@ -52,7 +52,7 @@ defmodule OMG.WatcherInfo.MixProject do
       {:briefly, "~> 0.3.0", only: [:dev, :test]},
       {:omg_child_chain, in_umbrella: true, only: [:test], runtime: false},
       {:phoenix, "~> 1.3", runtime: false},
-      {:ex_machina, "~> 2.3", only: [:test], runtime: false}
+      {:ex_machina, "~> 2.3", only: [:test, :dev], runtime: false}
```
Comment (unnawut) [inline apps/omg_watcher_info/mix.exs]: Can discard this change now

File: apps/omg_db/test/omg_db/release_tasks/set_key_value_db_test.exs (hunk 8 above)
Comment (InoMurko) [reply apps/omg_db/test/omg_db/release_tasks/set_key_value_db_test.exs]: because of the DB path: https://github.com/omisego/elixir-omg/pull/1027/files#diff-2fb51e4c80ccbd9db57164cb0023d00aR44

if you run things barebone locally, watcher and watcher_inf would point to the same directory

File: apps/omg_child_chain/lib/omg_child_chain/child_manager.ex (hunk 6 above)
Comment (InoMurko) [reply apps/omg_child_chain/lib/omg_child_chain/child_manager.ex]: no technical reason! reduced it to 100miliseconds. I would expand the manager if we see it beneficial, so that processes on the same level checkin to him and once they're all checked it, it would forward the checkin to monitor. for now it works OK.

File: apps/omg_child_chain/test/omg_child_chain/integration/fixtures.exs (hunk 7 above)
Comment (InoMurko) [reply apps/omg_child_chain/test/omg_child_chain/integration/fixtures.exs]: added 10seconds

File: apps/omg_child_chain/test/omg_child_chain/integration/happy_path_test.exs [hunk 11]
```diff
@@ -263,7 +263,9 @@ defmodule OMG.ChildChain.Integration.HappyPathTest do
 
   defp get_input_txs(txs), do: Enum.map(txs, &Transaction.raw_txbytes/1)
 
-  defp wait_for_web() do
+  defp wait_for_web(), do: wait_for_web(100)
```
Comment (unnawut) [inline apps/omg_child_chain/test/omg_child_chain/integration/happy_path_test.exs]: Missing `wait_for_web(counter - 1)` a few lines below

File: apps/omg_watcher/test/fixtures.exs [hunk 12]
```diff
@@ -200,7 +200,9 @@ defmodule OMG.Watcher.Fixtures do
     }
   end
 
-  defp wait_for_web() do
+  defp wait_for_web(), do: wait_for_web(100)
```
Comment (unnawut) [inline apps/omg_watcher/test/fixtures.exs]: Missing `wait_for_web(counter - 1)` a few lines below

File: apps/omg_child_chain/test/omg_child_chain/integration/fixtures.exs [hunk 13]
```diff
@@ -96,4 +98,17 @@ defmodule OMG.ChildChain.Integration.Fixtures do
...
     {deposit_blknum, token_deposit_blknum}
   end
+
+  defp wait_for_web(), do: wait_for_web(100)
+
+  defp wait_for_web(counter) do
+    case Keyword.has_key?(Alarm.all(), elem(Alarm.main_supervisor_halted(__MODULE__), 0)) do
+      true ->
+        Process.sleep(100)
+        wait_for_web(counter(-1))
```
Comment (unnawut) [inline apps/omg_child_chain/test/omg_child_chain/integration/fixtures.exs]:
```suggestion
        wait_for_web(counter - 1)
```
